ajax/itilfollowup.php

```diff
@@ -44,16 +44,42 @@
 
 Session::checkLoginUser();
 
-if (isset($_POST['itilfollowuptemplates_id'])
-    && $_POST['itilfollowuptemplates_id'] > 0) {
-   $template = new ITILFollowupTemplate();
-   $template->getFromDB($_POST['itilfollowuptemplates_id']);
-
-   echo json_encode(
-      [
-         'content'         => RichText::getSafeHtml($template->fields['content'], true),
-         'requesttypes_id' => $template->fields['requesttypes_id'],
-         'is_private'      => $template->fields['is_private'],
-      ]
-   );
+// Mandatory parameter: itilfollowuptemplates_id
+$itilfollowuptemplates_id = $_POST['itilfollowuptemplates_id'] ?? 0;
+if (!$itilfollowuptemplates_id) {
+   Toolbox::throwError(400, "Missing or invalid parameter: 'itilfollowuptemplates_id'");
 }
+
+// Mandatory parameter: parents_id
+$parents_id = $_POST['parents_id'] ?? 0;
+if (!$parents_id) {
+   Toolbox::throwError(400, "Missing or invalid parameter: 'parents_id'");
+}
+
+// Mandatory parameter: parents_itemtype
+$parents_itemtype = $_POST['parents_itemtype'] ?? '';
+if (empty($parents_itemtype) || !is_subclass_of($parents_itemtype, CommonITILObject::class)) {
+   Toolbox::throwError(400, "Missing or invalid parameter: 'parents_itemtype'");
+}
+
+// Load followup template
+$template = new ITILFollowupTemplate();
+if (!$template->getFromDB($itilfollowuptemplates_id)) {
+   Toolbox::throwError(400, "Unable to load template: $itilfollowuptemplates_id");
+}
+
+// Load parent item
+$parent = new $parents_itemtype();
+if (!$parent->getFromDB($parents_id)) {
+   Toolbox::throwError(400, "Unable to load parent item: $parents_itemtype $parents_id");
+}
+
+// Render template content using twig
+$parameters_class = $parent::getUserTemplatesParametersClass();
+$parameters = new $parameters_class();
+$template->fields['content'] = UserTemplates::render($template->fields['content'], [
+   'ticket' => $parameters->getValues($parent)
+]);
```

Comment:
This is code taken from a controller when rendering the template, it shouldn't prevent you from adding more parameters like this:
```
$template->fields['content'] = UserTemplates::render($template->fields['content'], [
   'event' => $event,
   'ticket' => $parameters->getValues($parent),
]);
```
We do need to keep a 'ticket' => value key since the values in the template will be ticket.xxx .

Would this work for you ?

Reply:
Another solution could be to add a second parameter to `AbstractTemplatesParameters::getValues()`, and a new method `AbstractTemplatesParameters::getRootName()`, so that caller could indicates if values should be encapsulated in their "root" variable or not.

```
public function getRootName(): string {
   return 'ticket';
}
public function getValues(CommonDBTM $item, bool $root = false): array {
   ...
   $values = $this->defineValues($item);
   return $root ? [$this->getRootName() => $values] : $values;
}
```

With this logic, root variable name will be defined by the parameters class itself, limiting the potential for error when copying / pasting.

Reply:
Done 3e49dafc38.